src/ILCompiler.ReadyToRun/src/Compiler/DependencyAnalysis/ReadyToRunCodegenNodeFactory.cs

```diff
@@ -767,25 +768,54 @@ public ISymbolNode DelegateCtor(TypeDesc delegateType, MethodDesc targetMethod,
             TypeAndMethod ctorKey = new TypeAndMethod(delegateType, targetMethod);
             if (!_delegateCtors.TryGetValue(ctorKey, out ctorNode))
             {
-                ModuleToken delegateTypeToken;
-                if (CompilationModuleGroup.ContainsType(delegateType) && delegateType is EcmaType ecmaType)
-                {
-                    delegateTypeToken = new ModuleToken(ecmaType.EcmaModule, (mdToken)MetadataTokens.GetToken(ecmaType.Handle));
-                }
-                else
-                {
-                    // TODO: reverse typedef lookup within the version bubble
-                    throw new NotImplementedException();
-                }
-
                 IMethodNode targetMethodNode = MethodEntrypoint(targetMethod, methodToken, isUnboxingStub: false);
 
                 ctorNode = new DelayLoadHelperImport(this,
                     ILCompiler.DependencyAnalysis.ReadyToRun.ReadyToRunHelper.READYTORUN_HELPER_DelayLoad_Helper,
-                    new DelegateCtorSignature(this, delegateType, delegateTypeToken, targetMethodNode, methodToken));
+                    new DelegateCtorSignature(this, delegateType, default(ModuleToken), targetMethodNode, methodToken));
                 _delegateCtors.Add(ctorKey, ctorNode);
             }
             return ctorNode;
         }
+
+        /// <summary>
+        /// Reverse lookup table mapping external types to reference tokens in the input modules. The table
+        /// gets lazily initialized with all input module references the first time the reverse lookup is
+        /// needed.
+        /// </summary>
+        private Dictionary<TypeDesc, ModuleToken> _typeToRefTokens = null;
```

Comment:
Wouldn't it be easier to keep track of this in the JitInterface? RyuJIT is going to call `resolveToken` to resolve all tokens. We could just stash away those that have `IsDelegate == true` and when JitInterface calls `NodeFactory.DelegateCtor`, we could just find the token in the map and provide it as a parameter there. The fix would be more localized to the place that has the problem.

That way we avoid having to go over all the TypeRef tokens that is just asking for extra trouble (users love to ship their code with incomplete/unresolvable/junk type references).

Reply:
Hmm, I just tried this out but unfortunately it doesn't work. We don't only need the tokens for delegate types, we need them in all sorts of places around the signature encoding when we don't have a token readily available which is the case most of the time for composite signatures like instantiations.

In particular, just using pre-recorded calls to resolveToken I'm unable to resolve List`1 because that never enters resolveToken by itself even though many methods on an instantiation of the type do.

Gracefully ignoring unresolvable references is probably a worthy goal; please let me also remind you that in the expected typical user case the "external references" involve "our" framework code. Other than that, the current implementation is more or less an improved version of

https://github.com/dotnet/coreclr/blob/61146b5c5851698e113e936d4e4b51b628095f27/src/vm/ceeload.cpp#L5872